### DISCUSSION TITLE
Realm Collection

### DISCUSSION THREAD

**Discussion by jbtdevgit** [realm/realm-dotnet#3465] [Q&A] [answered]
Hi all,

I am quite stumped on how to properly manage a collection when I am adding, updating, or deleting to a collection. Do I need to open the realm whenever I want to modify a collection?  Whenever a collection is fetched, does the realm need to be open at all times (except when exiting/removing pages)?

Right now I have a `DatabaseService` that implements my crud, so I open and close my `realm` instances whenever I access or have to do something in related to the database. I also do my adding and editing of an item is in a separate page so subscribing to realm collection does not make sense for me to do so. I am not sure if it is a plausible solution to fetch items and refresh a page again is a better solution than adding the newly added item at the end of a collection.

Thanks for any clarification.

**Comment by nirinchev** [comment, marked as answer]
Realm objects and collections are not copied into memory - instead data is read directly from the database when you read properties/indexes from them and is written directly to the database when you mutate them. This means that, yes, the Realm that owns an object/collection needs to be kept open for the lifetime of the object/collection.

Generally, Realm is not a great fit for a repository pattern, because it relies heavily on lazy-loading and reactivity. You can achieve it by copying the data from the database into unmanaged objects either manually or using automapper, but doing so will lose out on a lot of the benefits of using Realm.

**Reply by jbtdevgit** [reply]
I see, I've done the ORM and yes, you are correct with that, I've had problems with updates/deletes of realm object as they become unmanaged objects the moment I close the realm. 

So to summarise, realm cannot be opened and closed at given instances, not until I am done with the object itself. Is that correct?

**Reply by papafe** [reply]
Yes, you should keep the realm open until you need the objects. 
It obviously depends on your kind of application, but you could keep a realm opened for the whole lifetime of your app. A pattern that we generally recommend is:
- Have one realm opened on the main thread of your application that you use most of the time.
- Eventually temporary realms that you open on other threads as needed. Remember to dispose of those when you don't need them anymore, otherwise you could have some performance issue.
